Noaaweather

rewsssiestedbo opened the pull request:
## Description:

Add sensor to get weather measurements from US NOAA/National Weather Service

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>
Not yet submitted

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: noaaweather
    name: Kailua NWS conditions
    latitude: 21.4131155
    longitude: -157.7573696
    monitored_conditions:
        - temperature
        - textDescription
        - dewpoint
        - windChill
        - windSpeed
        - windDirection
        - windGust
        - barometricPressure
        - seaLevelPressure
        - precipitationLastHour
        - relativeHumidity
        - visibility
        - cloudLayers

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

rewsssiestedbo committed:
Initial commit of noaaweather sensor

rewsssiestedbo committed:
clean up unit conversion

rewsssiestedbo committed:
Additional clean up to meet coding standards.

rewsssiestedbo committed:
Set up test cases

rewsssiestedbo committed:
clean up formating for test code

rewsssiestedbo committed:
clean up formating for test code

rewsssiestedbo committed:
clean up formating for test code

rewsssiestedbo committed:
error handling and test setup

rewsssiestedbo committed:
Check qualityControl values to identify provided data

rewsssiestedbo committed:
Check qualityControl values to identify provided data

rewsssiestedbo committed:
Use individual timestamps per condition to allow tracking of updates without values

rewsssiestedbo committed:
Get all tests running with async configuration

rewsssiestedbo committed:
Clarify some comments

rewsssiestedbo committed:
Clarify some comments, clean up some formatting

rewsssiestedbo committed:
Use constants for various units

rewsssiestedbo committed:
When METAR data (rawMessage) is available, use values from the METAR data for missing values in normal JSON attributes

rewsssiestedbo committed:
Add tests for METAR record usage, fix issues found

rewsssiestedbo committed:
Set data update time when using METAR value, add special case for missing windGust value

rewsssiestedbo committed:
Add translation of weather description to homeassistant descriptions

rewsssiestedbo committed:
Finish optional conversion of NOAA/NWS weather descriptions to use
Home Assistant weather condition terms.  Include check for whether
it is night at the location.

rewsssiestedbo committed:
Fix checks for night by estimating time offset for location

rewsssiestedbo committed:
Fix casing of text descriptions of measurements

rewsssiestedbo committed:
Update to version 0.6 of pynws

rewsssiestedbo committed:
Run script.gen_requirements_all and script.hassfest

homeassistant commented:
Hello @rewsssiestedbo,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`9b1a56f45a434a4f5a768fb37e0fd74467a1f15e`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;23875&#x2F;commits&#x2F;9b1a56f45a434a4f5a768fb37e0fd74467a1f15e) - This commit has something that looks like an email address (hadev@bush03.localhost). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hello @rewsssiestedbo,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`9b1a56f45a434a4f5a768fb37e0fd74467a1f15e`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;23875&#x2F;commits&#x2F;9b1a56f45a434a4f5a768fb37e0fd74467a1f15e) - This commit has something that looks like an email address (hadev@bush03.localhost). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hi @rewsssiestedbo, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#23875).

Once you do that we will be able to review and accept this pull request.

Thanks!

rewsssiestedbo committed:
Add noaaweather requirements to gen_requirements_all.py and run script to
generate requirements_test_all.txt

MatthewFlamm commented:
This needs to be considered along with the weather platform PR in [#23647](https://github.com/home-assistant/home-assistant/pull/23647).  I think there should be consistent naming and methodology.  I propose NWS since it the sub-agency of NOAA that provides the data.  I have no strong position here however.

There are also several different methods that were chosen in our approaches.  For example, here, the HA condition is derived through some rules from the long text description while the weather platform PR uses the icon url provided in the nws data.  These will sometimes give inconsistent results.  I like the icon url approach as it is straightforward, but I'd like to hear if there is a benefit to your approach that I haven't considered.

Another example is that you use the raw metar code as the fallback for missing or low quality data.  The weather platform PR does not use a fall back at all currently and in fact will use data marked as poor quality.  I planned on falling back to the previous observation in this case, to be introduced in a future PR.  We should align on a common fallback methodology.

I propose that we work together to iron out these differences, starting with the above two points, but I will read your PR in more detail.

rewsssiestedbo committed:
Fix typo in string (missing "n" in "lightning")

rewsssiestedbo committed:
Add tests for some additional error conditions.

rewsssiestedbo commented:
> This needs to be considered along with the weather platform PR in [#23647](https://github.com/home-assistant/home-assistant/pull/23647). I think there should be consistent naming and methodology. I propose NWS since it the sub-agency of NOAA that provides the data. I have no strong position here however.
> 
> There are also several different methods that were chosen in our approaches. For example, here, the HA condition is derived through some rules from the long text description while the weather platform PR uses the icon url provided in the nws data. These will sometimes give inconsistent results. I like the icon url approach as it is straightforward, but I'd like to hear if there is a benefit to your approach that I haven't considered.
> 
> Another example is that you use the raw metar code as the fallback for missing or low quality data. The weather platform PR does not use a fall back at all currently and in fact will use data marked as poor quality. I planned on falling back to the previous observation in this case, to be introduced in a future PR. We should align on a common fallback methodology.
> 
> I propose that we work together to iron out these differences, starting with the above two points, but I will read your PR in more detail.

Yes, we should work together on this.   I have no particular concern about using nws instead of noaa,  If I recall, the only reason I went with noaa it seemed less likely to have a conflict with some other countries weather service abbreviation, and there was already the NOAA Tides sensor.
As far as using the icon URLs versus parsing the text description, it was not clear to me whether the icon URLs are at all stable long term.   Overall the documentation on the observation values returned is fairly sparse, so it is not clear whether they have any commitment to keeping the icon URLs stable.   The text description values have been stable for a long time, but not guaranteed either.   If my code is coming up with different results from yours, I may have just interpreted something differently or made a different arbitrary choice:  there is certainly not a one to one mapping to the HA condition values.

Finally, about dealing with missing information in responses.  I initially just had the code to avoid updating a condition if the response was missing a value for that condition.   However, for a few stations I was monitoring (PHNG and PHNL show this), there would be long periods with no valid temperature value (for example) in the temperature field, but with an updated value in the METAR record.   It is not clear to me why this happens, but it is not unusual for there to be multiple observation records without valid values for a number of conditions, but with valid values in the METAR record.

I'll try to take a look are your PR and see how best to align.

rewsssiestedbo committed:
Remove extra check for no stations returned on request for station list

rewsssiestedbo committed:
fix formating and ensure one bad METAR record in data

MatthewFlamm commented:
I've noticed too that sometimes some stations are better than others.  The closest station to me is always sunny or cloudy, but never rainy.  My nearby stations will only be missing/contain poor quality data for the current observation, however.

I think your metar fall back is the right approach for your case, but adds quite the complexity to the HA code base.  It seems to me like it would be better to have these decisions unified in one module(or a new pypi library) rather than spread across two platforms.  I need to think about the path forward, but one suggestion is to co-create a new pypi library that does all the data cleansing and presents HA with a very clean API.

For the icon, the big upside I see is that they publish a list of icons.  So if they change the icon names and they continue to publish this list, it will be easy to maintain.  Your method could be a nightmare if it ever changes.  You'd have to deduce the new structure of the text.

rewsssiestedbo commented:
> I've noticed too that sometimes some stations are better than others. The closest station to me is always sunny or cloudy, but never rainy. My nearby stations will only be missing/contain poor quality data for the current observation, however.
I suspect that the closest station to you does not have a rain gauge.   There are clearly a number of different types of stations:  those at airports that are fully staffed with weather related personal and full sets of instruments; fully automated stations at smaller airports, which may or may not have a full set of instruments; stations run by organizations other than the NWS, which only have the instruments they care about.  An example of the last case is GGWPW, which appears to have temperature, dewpoint, wind speed and direction and some precipitation detector.  It appears to have no cloud detection, so it will never return cloudy, and I'm not sure if it ever gives anything other then Clear plus a wind indication.
> 
> I think your metar fall back is the right approach for your case, but adds quite the complexity to the HA code base. It seems to me like it would be better to have these decisions unified in one module(or a new pypi library) rather than spread across two platforms. I need to think about the path forward, but one suggestion is to co-create a new pypi library that does all the data cleansing and presents HA with a very clean API.
I tend to agree with you here.   When I started I assumed that I could simply take the values and deal with unit conversion and be done.  The complexity that is there now is a result of the incremental increases that came from realizing how messy the data really is.  Creating an independent modules for getting the observation data and dealing with the mess can simplify the HA specific code enormously.  
Given the organization of the NWS APIs, it would probably be best to have an observation station object model and an independent forecast object model, since the observations are dependent upon particular stations, while the forecast are for a specific grid point area, and they update at different rates.
> 
> For the icon, the big upside I see is that they publish a list of icons. So if they change the icon names and they continue to publish this list, it will be easy to maintain. Your method could be a nightmare if it ever changes. You'd have to deduce the new structure of the text.

Note that the translation to the HA weather code values was the last thing I added, and is only used as an option, since the NWS text descriptions have much more content than the HA codes, and I would rather have the direct NWS text.

Any thoughts on how to get started on creating the separate (from HA) module(s)?  

For the forecast data, had you looked at using the NDFD API instead of the one you are using?   It would be somewhat more complex to handle the XML responses, but it provides much more data, including precipitation amounts, dewpoint, cloud cover, etc.

MatthewFlamm commented:
After reading through your PR again, it seems like the inconsistencies are mainly the ones I already mentioned, so I don't think we need a separate library for this.  The biggest decision is to coalesce on the name in my mind.  I will look into implementing a metar fallback in my PR.  Do you have a reference for the qc code info you used?



> For the forecast data, had you looked at using the NDFD API instead of the one you are using? It would be somewhat more complex to handle the XML responses, but it provides much more data, including precipitation amounts, dewpoint, cloud cover, etc.

I did look at it, but it again adds huge complexity.  The weather.gov API is very simple and clean.

rewsssiestedbo committed:
Remove checks for no unit values in unit_of_measure property, since
this should have been handled in the update code already.

Add HA weather condition of "hail".  While this is not in the documentation
it is in the demo component.

rewsssiestedbo committed:
Add some additional test cases to cover more error conditions.

rewsssiestedbo commented:
> After reading through your PR again, it seems like the inconsistencies are mainly the ones I already mentioned, so I don't think we need a separate library for this. The biggest decision is to coalesce on the name in my mind. I will look into implementing a metar fallback in my PR. Do you have a reference for the qc code info you used?

I was never able to find any documentation on the qc codes.  The comments in sensors.py are a result of reverse engineering/review of the results from a number of different stations over a few months. 

For the HA conditions, it appears that the differences in the results from using the icons versus the text are for:

1. "lightning" versus "lightning-rainy", since I chose to have "Thunderstorm in vicinity" be treated as "lightning", and other values as "lightning-rainy".
2. "pouring" versus "rainy", since the NWS icons for rain and heavy rain are the same, but the text is different.
3. "hail" - I just added this, since it is not in the documentation page for a weather component, but is in the demo code.

As far as the name, it makes sense to use nws.

> 
> > For the forecast data, had you looked at using the NDFD API instead of the one you are using? It would be somewhat more complex to handle the XML responses, but it provides much more data, including precipitation amounts, dewpoint, cloud cover, etc.
> 
> I did look at it, but it again adds huge complexity. The weather.gov API is very simple and clean.

rewsssiestedbo committed:
Correct entry for HA weather condition snowy

rewsssiestedbo committed:
Set some observation records to ensure tests flow through certain paths of the code
that depend upon observation data:  either the time of day or the values available.

rewsssiestedbo committed:
Add checks for bad JSON responses

rewsssiestedbo committed:
Add test fixtures for bad JSON responses, missing station list

rewsssiestedbo committed:
Remove some redundant error checks

rewsssiestedbo committed:
Update test data to include specific cases of missing or bad data.
Use specific data for night time check.  Note this also tests
duplicate observation timestamps.

rewsssiestedbo committed:
Change bad json test file to be more typical failure - file cut short instead of embedded error

rewsssiestedbo committed:
Correct timestamps to ensure duplicates

rewsssiestedbo committed:
Correct special case processing of missing windGust value
Update timestamps in test data to ensure duplicate timestamps.

MatthewFlamm reviewed (commented):
Some comments

MatthewFlamm commented on homeassistant/components/noaaweather/sensor.py:
```diff
@@ -0,0 +1,1147 @@
+    #
+    nws = pynws.Nws(session, latlon=(latitude, longitude))
+    nws.userid = config.get(CONF_USERAGENT)
+    nws.limit = 5
```
The limit attribute that is

rewsssiestedbo replied:
True - I missed that change in pwnws 0.6.  I moved the limit=5 to the call to observations.

rewsssiestedbo committed:
Correct use of limit for getting observations to match pynws 0.6 behavior

rewsssiestedbo committed:
use homeassistant.util conversions in unit_convert when there is one available

rewsssiestedbo committed:
Remove trailing space

rewsssiestedbo committed:
Remove usehaweather option - always just provide the textDescription value as is.

rewsssiestedbo committed:
Remove timestamp and rawMessage from one observation to ensure test of code handling those items missing

rewsssiestedbo committed:
Clean up test case comments

rewsssiestedbo committed:
Remove checks for NwsError, since pynws will only raise that exception when bad parameters are given

rewsssiestedbo commented on homeassistant/components/noaaweather/sensor.py:
```diff
@@ -0,0 +1,1147 @@
+    """Convert the measurement from the native unit to the appropriate unit.
+
+    This will do the appropriate conversion to get the value
+    from unit it is supplied in to the standard metric or imperial
```
Changed to use the built in unit conversions for temperature and pressure.  As the distance one does not have inches for millimeters, and there are no speed conversions, I left those as is.  In a future pull request I may try to add the builtin conversion for speed and the additional distance units.

rewsssiestedbo committed:
Include record with variable wind direction

MatthewFlamm commented:
I added the metar fallback logic into the upstream `pynws` library with a new class. My implementation was simply to check for missing values in the observation.  For missing values, replace with metar values, if available.  I only put in the metar fallback for the weather platform variables (temperature, pressure, etc.) so far, but it is written in a way that should be straightforward for other variables if they are contained in the metar report and can be parsed with the metar library.

I'm willing to consider more complex logic in the pynws library, too if warranted.

rewsssiestedbo commented:
Closing to wait for integration of nws weather component, and then integration of the sensor code with that component.
